lte/gateway/c/session_manager/LocalEnforcer.cpp

```diff
@@ -408,29 +408,53 @@ void LocalEnforcer::aggregate_records(
 
   for (const RuleRecord& record : records.records()) {
     const std::string &imsi = record.sid(), &ip = record.ue_ipv4();
+    const uint32_t teid = record.teid();
     // TODO IPv6 add ipv6 to search criteria
     SessionSearchCriteria criteria(imsi, IMSI_AND_UE_IPV4_OR_IPV6, ip);
     auto session_it = session_store_.find_session(session_map, criteria);
     if (!session_it) {
-      MLOG(MERROR) << "Could not find an active session for " << imsi << " and "
-                   << ip << " during record aggregation";
-      dead_sessions_to_cleanup.insert(record);
-      continue;
-    }
+      MLOG(MERROR) << "Could not find an active session for 4G " << imsi
+                   << " and " << ip << " during record aggregation";
+      if (teid) {
+        SessionSearchCriteria criteria(imsi, IMSI_AND_TEID, teid);
+        auto session_it = session_store_.find_session(session_map, criteria);
+        if (!session_it) {
+          dead_sessions_to_cleanup.insert(record);
+          continue;
+        }
+        auto& session                 = **session_it;
+        const std::string& session_id = session->get_session_id();
+        sessions_with_reporting_flows.insert(
+            ImsiAndSessionID(imsi, session_id));
+        if (record.bytes_tx() > 0 || record.bytes_rx() > 0) {
+          MLOG(MDEBUG) << session_id << " used " << record.bytes_tx()
+                       << " tx bytes and " << record.bytes_rx()
+                       << " rx bytes for rule " << record.rule_id();
+        }
+        session->add_rule_usage(
+            record.rule_id(), record.rule_version(), record.bytes_tx(),
+            record.bytes_rx(), record.dropped_tx(), record.dropped_rx(),
+            &session_update[imsi][session_id]);
+        continue;
+      } else {
```

Comment:
You are duplicationg code here again. Remember that if you see yourself copy patsing then you probably can do it differentlty.
 
 In this function you just need to find an existing session (either 4g or 5g). Once you have the session the logic to apply to the session is the same.

 You have several options here
 - You can extend the `criteria` object to take two inpupts
 ```
criteria(imsi, IMSI_AND_UE_IPV4V6_OR_5GTEID, teid);
 ```
 - You can extend the session_store_.find_session method to take more than one criteria
 - You can make a separate function that will search using both criterias and will return session_it

Reply:
Yes, I agreed your point. Only my intention was not to disturb 4g session search. Now I have modified as you mentioned.